---
title: Practice with Python
author: Brenna Briles
date: '2019-12-15'
slug: practice-with-python
categories: []
tags: []
description: ''
---



<p>This post contains sample problems using Python from homework assignments in SDS348: Computational Biology and Bioinformatics instructed by Nathaniel Woodward and taken at the University of Texas at Austin.</p>
<div id="numpy" class="section level1">
<h1>NumPy</h1>
<p>Load the dataset faithful. Import numpy and use functions from it to compute the mean, median, and variance for each variable. Compute the correlation between the two variables using corrcoef.</p>
<pre class="python"><code>import csv
eruptions = []
waiting = []
with open(&#39;./faithful.csv&#39;, &#39;r&#39;) as f:
    f.readline()
    reader = csv.reader(f, delimiter=&#39;,&#39;)
    for row in reader:
        eruptions.append(float(row[1]))
        waiting.append(int(row[2]))</code></pre>
<pre><code>## &#39;&quot;&quot;,&quot;eruptions&quot;,&quot;waiting&quot;\n&#39;</code></pre>
<pre class="python"><code>import numpy as np
print(np.mean(eruptions))</code></pre>
<pre><code>## 3.4877830882352936</code></pre>
<pre class="python"><code>print(np.mean(waiting))</code></pre>
<pre><code>## 70.8970588235294</code></pre>
<pre class="python"><code>print(np.median(eruptions))</code></pre>
<pre><code>## 4.0</code></pre>
<pre class="python"><code>print(np.median(waiting))</code></pre>
<pre><code>## 76.0</code></pre>
<pre class="python"><code>print(np.var(eruptions))</code></pre>
<pre><code>## 1.2979388904492861</code></pre>
<pre class="python"><code>print(np.var(waiting))</code></pre>
<pre><code>## 184.14381487889273</code></pre>
<pre class="python"><code>print(np.corrcoef(eruptions, waiting))</code></pre>
<pre><code>## [[1.         0.90081117]
##  [0.90081117 1.        ]]</code></pre>
</div>
<div id="regex" class="section level1">
<h1>RegEx</h1>
<p>Using re.findall(), match all latex math-mode text (anything wrapped in $, including the $s) in the string below (string4).</p>
<pre class="python"><code>import re
string4=r&quot;We try to quantitatively capture these characteristics by defining a set of indexes,\
which can be computed using the mosaic image and the corresponding ground truth: \
\begin{itemize} \
    \item $\mu_{A_T}$ and $\sigma_{A_T}$, the mean and standard deviation of the tiles area $A_T$, respectively; \
    \item $\rho_\text{filler}$, the ratio between the filler area and the overall mosaic are, computed as \
$\rho_\text{filler}=\frac{\sum_{T \in \mathcal{T} A_T}}{A}$, being $A$ the area of the mosaic; \
    \item \todo{does it worth?}; \
    \item \todo{does it worth?}; \
    \item $\mu_{C_T}$, the mean of the tiles \emph{color dispersion} $C_T$, \
being $C_T = \sigma_R+\sigma_G+\sigma_B$, where $\sigma_R$, $\sigma_G$ and $\sigma_B$ are the \
standard deviation of the red, green and blue channel values of the pixels within the tile $T$.\
After applying a method to an image, we compare the segmented image (i.e., the result) \
against the ground truth and assess the performance according to the following three metrics: \
\begin{itemize} \
    \item average tile precision $P$ \
    \item average tile recall $R$ \
    \item tile count error $C$&quot;

print(re.findall(r&quot;\$.*?\$&quot;, string4) )</code></pre>
<pre><code>## [&#39;$\\mu_{A_T}$&#39;, &#39;$\\sigma_{A_T}$&#39;, &#39;$A_T$&#39;, &#39;$\\rho_\\text{filler}$&#39;, &#39;$\\rho_\\text{filler}=\\frac{\\sum_{T \\in \\mathcal{T} A_T}}{A}$&#39;, &#39;$A$&#39;, &#39;$\\mu_{C_T}$&#39;, &#39;$C_T$&#39;, &#39;$C_T = \\sigma_R+\\sigma_G+\\sigma_B$&#39;, &#39;$\\sigma_R$&#39;, &#39;$\\sigma_G$&#39;, &#39;$\\sigma_B$&#39;, &#39;$T$&#39;, &#39;$P$&#39;, &#39;$R$&#39;, &#39;$C$&#39;]</code></pre>
</div>
<div id="bioinformatics" class="section level1">
<h1>Bioinformatics</h1>
<p>In bioinformatics, k-mers refer to all the possible subsequences (of length k) from a read obtained through DNA sequencing. For example, if the DNA sequencing read is “ATCATCATG”, then a the 3-mers in that read include “ATC” (which occurs twice), “TCA” (which occurs twice), “CAT” (occurs twice), and “ATG” (occurs once). You can read more about k-mers on Wikipedia.</p>
<p>Write a function that takes a string of nucleotides as input and returns a dictionary with all 3-mers present in that string, and the number of times that each 3-mer occurs. Then use your function to find the 3-mers in the DNA sequence my_seq1 defined below.</p>
<pre class="python"><code>import re
my_seq1 = &quot;ATCATCATG&quot;
my_seq2 = &quot;CAGCCCAATCAGGCTCTACTGCCACTAAACTTACGCAGGATATATTTACGCCGACGTACT&quot;
def k_mer_freq(seq):
    code_freqs = {}
    for base_index in range(0, len(seq) - 2):
        cur_3_mer = str(seq[base_index:base_index+3])
        if (cur_3_mer in code_freqs):
            code_freqs[cur_3_mer] += 1
        else:
            code_freqs[cur_3_mer] = 1
    return code_freqs

print(k_mer_freq(my_seq1))</code></pre>
<pre><code>## {&#39;ATC&#39;: 2, &#39;TCA&#39;: 2, &#39;CAT&#39;: 2, &#39;ATG&#39;: 1}</code></pre>
<pre class="python"><code>print(k_mer_freq(my_seq2))</code></pre>
<pre><code>## {&#39;CAG&#39;: 3, &#39;AGC&#39;: 1, &#39;GCC&#39;: 3, &#39;CCC&#39;: 1, &#39;CCA&#39;: 2, &#39;CAA&#39;: 1, &#39;AAT&#39;: 1, &#39;ATC&#39;: 1, &#39;TCA&#39;: 1, &#39;AGG&#39;: 2, &#39;GGC&#39;: 1, &#39;GCT&#39;: 1, &#39;CTC&#39;: 1, &#39;TCT&#39;: 1, &#39;CTA&#39;: 2, &#39;TAC&#39;: 4, &#39;ACT&#39;: 4, &#39;CTG&#39;: 1, &#39;TGC&#39;: 1, &#39;CAC&#39;: 1, &#39;TAA&#39;: 1, &#39;AAA&#39;: 1, &#39;AAC&#39;: 1, &#39;CTT&#39;: 1, &#39;TTA&#39;: 2, &#39;ACG&#39;: 3, &#39;CGC&#39;: 2, &#39;GCA&#39;: 1, &#39;GGA&#39;: 1, &#39;GAT&#39;: 1, &#39;ATA&#39;: 2, &#39;TAT&#39;: 2, &#39;ATT&#39;: 1, &#39;TTT&#39;: 1, &#39;CCG&#39;: 1, &#39;CGA&#39;: 1, &#39;GAC&#39;: 1, &#39;CGT&#39;: 1, &#39;GTA&#39;: 1}</code></pre>
</div>
