---
title: Practice with Python
author: Brenna Briles
date: '2019-12-15'
slug: practice-with-python
categories: []
tags: []
description: ''
---
This post contains sample problems using Python from homework assignments in SDS348: Computational Biology and Bioinformatics instructed by Nathaniel Woodward and taken at the University of Texas at Austin.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reticulate)
use_python("/Library/Frameworks/Python.framework/Versions/3.8/bin/python3")
np <- import("numpy")

```
#NumPy

Load the dataset faithful. Import numpy and use functions from it to compute the mean, median, and variance for each variable. Compute the correlation between the two variables using corrcoef.

```{python}
import csv
eruptions = []
waiting = []
with open('./faithful.csv', 'r') as f:
    f.readline()
    reader = csv.reader(f, delimiter=',')
    for row in reader:
        eruptions.append(float(row[1]))
        waiting.append(int(row[2]))
import numpy as np
print(np.mean(eruptions))
print(np.mean(waiting))
print(np.median(eruptions))
print(np.median(waiting))
print(np.var(eruptions))
print(np.var(waiting))
print(np.corrcoef(eruptions, waiting))
```

#RegEx  

Using re.findall(), match all latex math-mode text (anything wrapped in $, including the $s) in the string below (string4).  
```{python}
import re
string4=r"We try to quantitatively capture these characteristics by defining a set of indexes,\
which can be computed using the mosaic image and the corresponding ground truth: \
\begin{itemize} \
    \item $\mu_{A_T}$ and $\sigma_{A_T}$, the mean and standard deviation of the tiles area $A_T$, respectively; \
    \item $\rho_\text{filler}$, the ratio between the filler area and the overall mosaic are, computed as \
$\rho_\text{filler}=\frac{\sum_{T \in \mathcal{T} A_T}}{A}$, being $A$ the area of the mosaic; \
    \item \todo{does it worth?}; \
    \item \todo{does it worth?}; \
    \item $\mu_{C_T}$, the mean of the tiles \emph{color dispersion} $C_T$, \
being $C_T = \sigma_R+\sigma_G+\sigma_B$, where $\sigma_R$, $\sigma_G$ and $\sigma_B$ are the \
standard deviation of the red, green and blue channel values of the pixels within the tile $T$.\
After applying a method to an image, we compare the segmented image (i.e., the result) \
against the ground truth and assess the performance according to the following three metrics: \
\begin{itemize} \
    \item average tile precision $P$ \
    \item average tile recall $R$ \
    \item tile count error $C$"

print(re.findall(r"\$.*?\$", string4) )
```

#Bioinformatics
In bioinformatics, k-mers refer to all the possible subsequences (of length k) from a read obtained through DNA sequencing. For example, if the DNA sequencing read is "ATCATCATG", then a the 3-mers in that read include "ATC" (which occurs twice), "TCA" (which occurs twice), "CAT" (occurs twice), and "ATG" (occurs once). You can read more about k-mers on Wikipedia.

Write a function that takes a string of nucleotides as input and returns a dictionary with all 3-mers present in that string, and the number of times that each 3-mer occurs. Then use your function to find the 3-mers in the DNA sequence my_seq1 defined below.
```{python}
import re
my_seq1 = "ATCATCATG"
my_seq2 = "CAGCCCAATCAGGCTCTACTGCCACTAAACTTACGCAGGATATATTTACGCCGACGTACT"
def k_mer_freq(seq):
    code_freqs = {}
    for base_index in range(0, len(seq) - 2):
        cur_3_mer = str(seq[base_index:base_index+3])
        if (cur_3_mer in code_freqs):
            code_freqs[cur_3_mer] += 1
        else:
            code_freqs[cur_3_mer] = 1
    return code_freqs

print(k_mer_freq(my_seq1))
print(k_mer_freq(my_seq2))
```